src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -136033,6 +136034,247 @@ partial class C2<T> : Base<
 
+        [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/40538")]
+        public void AnnotationMismatch_Interface_GenericClass_01()
+        {
+            var source = """
+                #nullable enable
+                var c = new C<int>();
+
+                partial class C<T> : I<T> { }
+
+                partial class C<T> :
+                    #nullable disable
+                    I<T>
+                    #nullable enable
+                {
+                }
+
+                public interface I<T> { }
+                """;
+
+            CompileAndVerify(CreateCompilationWithMscorlib40(source), symbolValidator: validate).VerifyDiagnostics();
+
+            static void validate(ModuleSymbol module)
+            {
+                var c = module.GlobalNamespace.GetTypeMember("C");
+                var cT = c.TypeParameters.Single();
+                AssertEx.Equal("T", cT.ToTestDisplayString(includeNonNullable: true));
+                Assert.Equal(CodeAnalysis.NullableAnnotation.None, cT.GetPublicSymbol().ReferenceTypeConstraintNullableAnnotation);
+
+                var i = c.AllInterfacesNoUseSiteDiagnostics.Single();
+                AssertEx.Equal("I<T>", i.ToTestDisplayString(includeNonNullable: true));
+            }
+        }
+
+        [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/40538")]
+        public void AnnotationMismatch_Interface_GenericClass_02()
+        {
+            var source = """
+                #nullable enable
+                var c = new C<int>();
+
+                partial class C<T> : I<T, T> { }
+
+                partial class C<T> : I<T,
+                    #nullable disable
+                    T>
+                    #nullable enable
+                {
+                }
+
+                public interface I<T1, T2> { }
+                """;
+            CompileAndVerify(source, symbolValidator: validate).VerifyDiagnostics();
+
+            static void validate(ModuleSymbol module)
+            {
+                var c = module.GlobalNamespace.GetTypeMember("C");
+                var cT = c.TypeParameters.Single();
+                AssertEx.Equal("T", cT.ToTestDisplayString(includeNonNullable: true));
+                Assert.Equal(CodeAnalysis.NullableAnnotation.None, cT.GetPublicSymbol().ReferenceTypeConstraintNullableAnnotation);
+
+                var i = c.AllInterfacesNoUseSiteDiagnostics.Single();
+                AssertEx.Equal("I<T, T>", i.ToTestDisplayString(includeNonNullable: true));
+            }
+        }
+
+        [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/40538")]
+        public void AnnotationMismatch_Interface_GenericClass_03()
+        {
+            var source = """
+                #nullable enable
+                var c1 = new C1<int>();
+                var c2 = new C2<int>();
+
+                partial class C1<T> : I<T, object?>;
+
+                partial class C1<T> : I<
+                    T,
+                #nullable disable
```

Comment:
I think it would also be interesting to test when the lexical first declaration has a nullable-disabled type argument